8Knot/pages/codebase/visualizations/cntrb_file_heatmap.py

```diff
@@ -0,0 +1,374 @@
+                ),
+            ]
+        )
+    ],
+)
+
+# callback for graph info popover
+@callback(
+    Output(f"popover-{PAGE}-{VIZ_ID}", "is_open"),
+    [Input(f"popover-target-{PAGE}-{VIZ_ID}", "n_clicks")],
+    [State(f"popover-{PAGE}-{VIZ_ID}", "is_open")],
+)
+def toggle_popover(n, is_open):
+    if n:
+        return not is_open
+    return is_open
+
+
+# callback for populating repo drop down
+@callback(
+    [
+        Output(f"repo-{PAGE}-{VIZ_ID}", "data"),
+        Output(f"repo-{PAGE}-{VIZ_ID}", "value"),
+    ],
+    [Input("repo-choices", "data")],
+)
+def repo_dropdown(repo_ids):
+
+    # array to hold repo_id and git url pairing for dropdown
+    data_array = []
+    for repo_id in repo_ids:
+        entry = {"value": repo_id, "label": augur.repo_id_to_git(repo_id)}
+        data_array.append(entry)
+    return data_array, repo_ids[0]
+
+
+# callback for populating directory drop down
+@callback(
+    [
+        Output(f"directory-{PAGE}-{VIZ_ID}", "data"),
+        Output(f"directory-{PAGE}-{VIZ_ID}", "value"),
+    ],
+    [Input(f"repo-{PAGE}-{VIZ_ID}", "value")],
+    background=True,
+)
+def directory_dropdown(repo_id):
+    # wait for data to asynchronously download and become available.
+    print(repo_id)
+    cache = cm()
+    df = cache.grabm(func=rfq, repos=[repo_id])
+    while df is None:
+        time.sleep(1.0)
+        df = cache.grabm(func=rfq, repos=[repo_id])
+
+    # test if there is data
+    if df.empty:
+        logging.warning(f"{VIZ_ID} DROPDOWN- NO DATA AVAILABLE")
+        return "NO DATA"
+
+    # strings to hold the values for each column (always the same for every row of this query)
+    repo_name = df["repo_name"].iloc[0]
+    repo_path = df["repo_path"].iloc[0]
+    repo_id = str(df["id"].iloc[0])
+
+    # pattern found in each file path, used to slice to get only the root file path
+    path_slice = repo_id + "-" + repo_path + "/" + repo_name + "/"
+    df["file_path"] = df["file_path"].str.rsplit(path_slice, n=1).str[1]
+
+    # drop columns not in the most recent collection
+    df = df[df["rl_analysis_date"] == df["rl_analysis_date"].max()]
+
+    # drop unneccessary columns not needed after preprocessing steps
+    df = df.reset_index()
+    df.drop(["index", "id", "repo_name", "repo_path", "rl_analysis_date"], axis=1, inplace=True)
+
+    # split file path by directory
+    df = df.join(df["file_path"].str.split("/", expand=True))
+
+    # take all of the files, split on the last instance of a / to get directories and top level files
+    directories = df["file_path"].str.rsplit("/", n=1).str[0].tolist()
+    directories = list(set(directories))
+
+    # get all of the file names to filter out of the directory set
+    top_level_files = df["file_name"][df[1].isnull()].tolist()
+    directories = [f for f in directories if f not in top_level_files]
+
+    # add top level directory to the list of directories
+    directories.append("Top Level Directory")
```

Comment:
sort the directory names so that sibling directories appear together.

Reply:
Good call, @sgoggins also mentioned putting top level directory at the beginning of the list. Ill make both of those changes, thinking about doing an alphabetical sort to get the sorting right

Reply:
++